.maintain/gitlab/check_companion_build.sh

```diff
@@ -0,0 +1,100 @@
+# pr like
+#
+# $REPO companion: $ORGANISATION/$REPO#567
+#
+# $ORGANISATION and $REPO are set using $1 and $2. You can also specify a custom
+# build command with $3
+# So invoke this script like:
+# ./check_companion_build.sh paritytech polkadot 'cargo test --release'
+
+set -e
+
+ORGANISATION=$1
+REPO=$2
+BUILDSTRING=${3:-cargo test --all --release}
+
+github_api_substrate_pull_url="https://api.github.com/repos/paritytech/substrate/pulls"
+# use github api v3 in order to access the data without authentication
+github_header="Authorization: token ${GITHUB_PR_TOKEN}"
+
+boldprint () { printf "|\n| \033[1m%s\033[0m\n|\n" "${@}"; }
+boldcat () { printf "|\n"; while read -r l; do printf "| \033[1m%s\033[0m\n" "${l}"; done; printf "|\n" ; }
+
+
+
+boldcat <<-EOT
+
+
+check_${REPO}_companion_build
+==============================
+
+this job checks if there is a string in the description of the pr like
+
+$REPO companion: $ORGANISATION/$REPO#567
+
+
+it will then run cargo check from this ${REPO}'s branch with substrate code
+from this pull request. otherwise, it will uses master instead
+
+
+EOT
+
+# Set the user name and email to make merging work
+git config --global user.name 'CI system'
+git config --global user.email '<>'
+
+# Merge master into our branch before building to make sure we don't miss
+# any commits that are required.
+git fetch --depth 100 origin
+git merge origin/master
+
+# Clone the current master branch into a local directory
+# NOTE: we need to pull enough commits to be able to find a common
+# ancestor for successfully performing merges below.
+git clone --depth 20 "https://github.com/${ORGANISATION}/${REPO}.git"
+
+cd "$REPO"
+
+# either it's a pull request then check for a companion otherwise use
+# master
+# shellcheck disable=SC2003
+if expr match "${CI_COMMIT_REF_NAME}" '^[0-9]\+$' >/dev/null
+then
+  boldprint "this is pull request no ${CI_COMMIT_REF_NAME}"
+
+  pr_data_file="$(mktemp)"
+  # get the last reference to a pr in the target repo
+  curl -sSL -H "${github_header}" -o "${pr_data_file}" \
+    "${github_api_substrate_pull_url}/${CI_COMMIT_REF_NAME}"
+
+  pr_body="$(sed -n -r 's/^[[:space:]]+"body": (".*")[^"]+$/\1/p' "${pr_data_file}")"
+
+  pr_companion="$(echo "${pr_body}" | sed -n -r \
+      -e "s;^.*[Cc]ompanion.*${ORGANISATION}/${REPO}#([0-9]+).*$;\1;p" \
+      -e "s;^.*[Cc]ompanion.*https://github.com/${ORGANISATION}/${REPO}/pull/([0-9]+).*$;\1;p" \
+    | tail -n 1)"
+
+  if [ "${pr_companion}" ]
+  then
+    boldprint "companion pr specified/detected: #${pr_companion}"
+    git fetch origin "refs/pull/${pr_companion}/head:pr/${pr_companion}"
+    git checkout "pr/${pr_companion}"
+    git merge origin/master
+  else
+    boldprint "no companion branch found - building ${REPO}:master"
+  fi
+  rm -f "${pr_data_file}"
+else
+  boldprint "this is not a pull request - building ${REPO}:master"
+fi
+
+# Patch all Substrate crates in the repo
+diener patch --crates-to-patch ../ --substrate --path Cargo.toml
```

Comment:
In the case of Cumulus, I think Polkadot has to be patched as well.

Reply:
It might be feasible to just check `Cargo.lock` for figuring out which repositories are dependent.

Take for instance https://github.com/paritytech/cumulus/blob/5eeb06c5839981a897ce3205bc0134e0bd136958/Cargo.lock#L748. That could be detected with something like the following:

```bash
declare -A match_arg
match_arg=()
match_arg["--substrate"]='source = "git+https://github.com/paritytech/substrate?'
match_arg["--polkadot"]='source = "git+https://github.com/paritytech/polkadot?'
match_arg["--beefy"]='source = "git+https://github.com/paritytech/parity-bridges-gadget?'

patch_args=()
while IFS= read -r cargo_lock; do
  while IFS= read -r line; do
    for patch_arg in "${!match_arg[@]}"; do
      if [ "${line:0:${#match_arg[$patch_arg]}}" = "${match_arg[$patch_arg]}" ]; then
        patch_args+=("$patch_arg")
        break 3
      fi
    done
  done < "$cargo_lock"
done < <(find **/Cargo.lock Cargo.lock)

diener patch --crates-to-patch ../ "${patch_args[@]}" --path Cargo.toml
```

Reply:
Cheers for that. Actually did it slightly differently (can't chain dependencies with diener afaik), but yeah. thanks for the guidance :)